README
NAME
    Net::Etcd - etcd v3 REST API.

SYNOPSIS
        Etcd v3.1.0 or greater is required.   To use the v3 API make sure to set environment
        variable ETCDCTL_API=3.  Precompiled binaries can be downloaded at https://github.com/etcd-io/etcd/releases.

        $etcd = Net::Etcd->new(); # host: 127.0.0.1 port: 2379
        $etcd = Net::Etcd->new({ host => $host, port => $port, ssl => 1 });

        # put key
        $put_key = $etcd->put({ key =>'foo1', value => 'bar' });

        # check for success of a transaction
        $put_key->is_success;

        # get single key
        $key = $etcd->range({ key =>'test0' });

        # return single key value or the first in a list.
        $key->get_value

        # get range of keys
        $range = $etcd->range({ key =>'test0', range_end => 'test100' });

        # return array { key => value } pairs from range request.
        my @users = $range->all

        # delete single key
        $etcd->deleterange({ key => 'test0' });

        # watch key range, streaming.
        $watch = $etcd->watch( { key => 'foo', range_end => 'fop'}, sub {
            my ($result) =  @_;
            print STDERR Dumper($result);
        })->create;

        # create/grant 20 second lease
        $etcd->lease( { ID => 7587821338341002662, TTL => 20 } )->grant;

        # attach lease to put
        $etcd->put( { key => 'foo2', value => 'bar2', lease => 7587821338341002662 } );

        # add new user
        $etcd->user( { name => 'samba', password => 'foo' } )->add;

        # add new user role
        $role = $etcd->role( { name => 'myrole' } )->add;

        # grant read permission for the foo key to myrole
        $etcd->role_perm( { name => 'myrole', key => 'foo', permType => 'READWRITE' } )->grant;

        # grant role
        $etcd->user_role( { user => 'samba', role => 'myrole' } )->grant;

        # defrag member's backend database
        $defrag = $etcd->maintenance()->defragment;
        print "Defrag request complete!" if $defrag->is_success;

        # member version
        $v = $etcd->version;

        # list members
        $etcd->member()->list;

DESCRIPTION
    Net::Etcd is object oriented interface to the v3 REST API provided by
    the etcd grpc-gateway <https://github.com/grpc-ecosystem/grpc-gateway>.

ACCESSORS
  host
    The etcd host. Defaults to 127.0.0.1

  port
    Default 2379.

  name
    Username for authentication, defaults to $ENV{ETCD_CLIENT_USERNAME}

  password
    Authentication credentials, defaults to $ENV{ETCD_CLIENT_PASSWORD}

  ca_file
    Path to ca_file, defaults to $ENV{ETCD_CLIENT_CA_FILE}

  key_file
    Path to key_file, defaults to $ENV{ETCD_CLIENT_KEY_FILE}

  cert_file
    Path to cert_file, defaults to $ENV{ETCD_CLIENT_CERT_FILE}

  cacert
    Path to cacert, defaults to $ENV{ETCD_CLIENT_CACERT_FILE}.

  ssl
    To enable set to 1

  api_version
    defaults to /v3beta

  api_path
    The full api path. Defaults to http://127.0.0.1:2379/v3alpha

  auth_token
    The token that is passed during authentication. This is generated during
    the authentication process and stored until no longer valid or username
    is changed.

PUBLIC METHODS
  version
    Returns the etcd member version

        $etcd->version()

  watch
    See Net::Etcd::Watch

        $etcd->watch({ key =>'foo', range_end => 'fop' })

  role
    See Net::Etcd::Auth::Role

        $etcd->role({ role => 'foo' });

  role_perm
    See Net::Etcd::Auth::RolePermission

    Grants or revoke permission of a specified key or range to a specified
    role.

        $etcd->role_perm(
            { name => 'myrole', key => 'bar', permType => 'READ', prefix => 1 } )->grant;

  user_role
    See Net::Etcd::User::Role

        $etcd->user_role({ name => 'samba', role => 'foo' });

  auth
    See Net::Etcd::Auth

        $etcd->auth({ name => 'samba', password => 'foo' })->authenticate;
        $etcd->auth()->enable;
        $etcd->auth()->disable

  lease
    See Net::Etcd::Lease

        $etcd->lease( { ID => 7587821338341002662, TTL => 20 } )->grant;

  maintenance
    See Net::Etcd::Maintenance

        $etcd->maintenance()->snapshot

  member
    See Net::Etcd::Member

        $etcd->member()->list

  user
    See Net::Etcd::User

        $etcd->user( { name => 'samba', password => 'foo' } )->add;

  put
    See Net::Etcd::KV::Put

        $etcd->put({ key =>'foo1', value => 'bar' });

  deleterange
    See Net::Etcd::KV::DeleteRange

        $etcd->deleterange({ key=>'test0' });

  range
    See Net::Etcd::KV::Range

        $etcd->range({ key =>'test0', range_end => 'test100' });

  txn
    See Net::Etcd::KV::Txn

        $etcd->txn({ compare => \@compare, success => \@op });

  op
    See Net::Etcd::KV::Op

        $etcd->op({ request_put => $put });
        $etcd->op({ request_delete_range => $range });

  compare
    See Net::Etcd::KV::Compare

        $etcd->compare( { key => 'foo', result => 'EQUAL', target => 'VALUE', value => 'baz' });
        $etcd->compare( { key => 'foo', target => 'CREATE', result => 'NOT_EQUAL', create_revision => '2' });

  configuration
    Initialize configuration checks to see if etcd is installed locally.

AUTHOR
    Sam Batschelet (hexfusion)

CONTRIBUTORS
    Ananth Kavuri Michael Fung

ACKNOWLEDGEMENTS
    The etcd <https://github.com/etcd-io/etcd> developers and community.

CAVEATS
    The etcd <https://github.com/etcd-io/etcd> v3 API is in heavy development
    and can change at anytime please see api_reference_v3
    <https://github.com/etcd-io/etcd/blob/master/Documentation/dev-guide/api_
    reference_v3.md> for latest details.

    Authentication provided by this module will only work with etcd v3.3.0+

LICENSE AND COPYRIGHT
    Copyright 2018 Sam Batschelet (hexfusion).

    This program is free software; you can redistribute it and/or modify it
    under the terms of either: the GNU General Public License as published
    by the Free Software Foundation; or the Artistic License.

    See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for Net::Etcd

0.023
[ BUG FIX ]
* Fix a bug where the Bool type would fail under certain circumstances when setting keys => 1 for ttl.
While I could replicate, travis tests failed consistantly, we will resolve this better in the future.
Reference bless( do{\(my $o = 1)}, 'JSON::PP::Boolean' ) did not pass type constraint "Bool" (in $args->{"keys"})

0.022
[ ENHANCEMENTS ]
* Improve support for prefix inline with etcd API. Fixes GH #53 Thanks Michael Fung!

0.021
[ ENHANCEMENTS ]
* Add support for prefix => 'true'.
* Add better handing of exceptions from from_json.

0.020
[ ENHANCEMENTS ]
* Add support for full chain TLS. Fixes GH #42
* Updated travis tests to use etcd v3.3.3

[ BREAKING ]
* Default value for api_version is now /v3beta.

[ ANNOUNCEMENT ]
* Default value for api_version will change to /v3 on the release of etcd v3.4

0.019
[ BUG FIX ]
* Resolved issue where chunked data was not properly handled in non streaming calls and would truncate data when buffer was full. Fixed By GH #40 - Thanks Ananth Kavuri (foobargeez)

0.018
[ ENHANCEMENTS ]
* Added Member class and tests.
* Added ENV defaults for name 'ETCD_CLIENT_USERNAME', password 'ETCD_CLIENT_PASSWORD' and cacert 'ETCD_CERT_FILE' accessors.
* Updated TLS example.

0.017
[ ENHANCEMENTS ]
* Add TLS support
* Added cacert accessor.
* Updated Travis to setup TLS etcd and run all tests using TLS.

0.016
[ ENHANCEMENTS ]
* Added status, defragment, version and status methods to Maintenance class.
* Added ability to check $etcd->version
* Add is_success and content helper methods to Action role.
* Updated POD.

0.015
[ ENHANCEMENTS ]
* Add ability to retry_auth. This feature is very useful for refreshing stale tokens. Closes GH #26

0.014
[ BUG FIX ]
* Resolved bug which prevented range delete transaction from completing. Closed via GH #25 Thanks Ananth Kavuri!

[ ANNOUNCEMENT ]
* Added class Net::Etcd::KV::RangeDelete
* Range delete is now rangedelete method calling delete against Range is deprecated.

0.013
[ ENHANCEMENTS ]
* Add Net::Etcd::Auth::RolePermission class allowing roles to be defined by key/range.

0.012
[ ENHANCEMENTS ]
* Add full support for header based authentication via grpc-gateway.  Requires etcd 3.2.0+
* Improve POD

0.011
[ ENHANCEMENTS ]
* Add full support for Txn with tests.  Recommended for use with etcd 3.2.0+

0.010
[ ENHANCEMENTS ]
* Add intial support for snapshot

0.009
[ ENHANCEMENTS ]
 * Add intial support for Txn
 * Add changepw support for users to change password.

0.008
[ ANNOUNCEMENT ]
 * Namespace change Etcd3 is now Net::Etcd

0.007

[ ANNOUNCEMENT ]
  * Etcd3 has been superceded by Net::Etcd and will be removed from CPAN.

0.006
[ ENHANCEMENTS ]
  * Sweeping changes to usage some not backwards compatible, sorry.
  * Restructuring of classes to consolidate and reduce redundancy, see d323feb5c0.
  * Replaced HTTP::Tiny with AnyEvent::HTTP allowing async interaction with etcd.
  * Simplified the internal way $etcd object is initialized.  Replacing ->connect with ->new, also moving ::Client into Net::Etcd module.
  * Renamed _client accessor to etcd.
  * Make Net::Etcd::KV a Moo::Role allowing easier interaction with other KV classes.


[ BUG FIXES ]
  * Fixed Watch support

[ DOCUMENTATION ]
  * Consolidate POD
  * Add Watch examples.

0.005
[ ENHANCEMENTS ]
  * Add lease support and tests

0.004
[ ENHANCEMENTS ]
  * travis testing added
[ BUG FIXES ]
  * Fix issue with HTTP::Tiny and support for Perl version where post function was not available.


lib/Net/Etcd.pm
use utf8;

package Net::Etcd;

# ABSTRACT: Provide access to the etcd v3 API.

use strict;
use warnings;

use JSON;
use MIME::Base64;
use Types::Standard qw(Str Int Bool HashRef);

use Net::Etcd::Auth;
use Net::Etcd::Auth::RolePermission;
use Net::Etcd::Config;
use Net::Etcd::Watch;
use Net::Etcd::Lease;
use Net::Etcd::Maintenance;
use Net::Etcd::Member;
use Net::Etcd::User;

use Moo;
with('Net::Etcd::KV');
use namespace::clean;

=encoding utf8

=head1 NAME

Net::Etcd - etcd v3 REST API.

=cut

our $VERSION = '0.022';

=head1 SYNOPSIS

    Etcd v3.1.0 or greater is required.   To use the v3 API make sure to set environment
    variable ETCDCTL_API=3.  Precompiled binaries can be downloaded at https://github.com/etcd-io/etcd/releases.

    $etcd = Net::Etcd->new(); # host: 127.0.0.1 port: 2379
    $etcd = Net::Etcd->new({ host => $host, port => $port, ssl => 1 });

    # put key
    $put_key = $etcd->put({ key =>'foo1', value => 'bar' });

    # check for success of a transaction
    $put_key->is_success;

    # get single key
    $key = $etcd->range({ key =>'test0' });

    # return single key value or the first in a list.
    $key->get_value

    # get range of keys
    $range = $etcd->range({ key =>'test0', range_end => 'test100' });

    # return array { key => value } pairs from range request.
    my @users = $range->all

    # delete single key
    $etcd->deleterange({ key => 'test0' });

    # watch key range, streaming.
    $watch = $etcd->watch( { key => 'foo', range_end => 'fop'}, sub {
        my ($result) =  @_;
        print STDERR Dumper($result);
    })->create;

    # create/grant 20 second lease
    $etcd->lease( { ID => 7587821338341002662, TTL => 20 } )->grant;

    # attach lease to put
    $etcd->put( { key => 'foo2', value => 'bar2', lease => 7587821338341002662 } );

    # add new user
    $etcd->user( { name => 'samba', password => 'foo' } )->add;

    # add new user role
    $role = $etcd->role( { name => 'myrole' } )->add;

    # grant read permission for the foo key to myrole
    $etcd->role_perm( { name => 'myrole', key => 'foo', permType => 'READWRITE' } )->grant;

    # grant role
    $etcd->user_role( { user => 'samba', role => 'myrole' } )->grant;

    # defrag member's backend database
    $defrag = $etcd->maintenance()->defragment;
    print "Defrag request complete!" if $defrag->is_success;

    # member version
    $v = $etcd->version;

    # list members
    $etcd->member()->list;

=head1 DESCRIPTION

L<Net::Etcd> is object oriented interface to the v3 REST API provided by the etcd L<grpc-gateway|https://github.com/grpc-ecosystem/grpc-gateway>.

=head1 ACCESSORS

=head2 host

The etcd host. Defaults to 127.0.0.1

=cut

has host => (
    is      => 'ro',
    isa     => Str,
    default => '127.0.0.1'
);

=head2 port

Default 2379.

=cut

has port => (
    is      => 'ro',
    isa     => Int,
    default => '2379'
);

=head2 name

Username for authentication, defaults to $ENV{ETCD_CLIENT_USERNAME}

=cut

has name => (
    is      => 'ro',
    default => $ENV{ETCD_CLIENT_USERNAME}
);

=head2 password

Authentication credentials, defaults to $ENV{ETCD_CLIENT_PASSWORD}

=cut

has password => (
    is      => 'ro',
    default => $ENV{ETCD_CLIENT_PASSWORD}
);

=head2 ca_file

Path to ca_file, defaults to $ENV{ETCD_CLIENT_CA_FILE}

=cut

has ca_file => (
    is      => 'ro',
    default => $ENV{ETCD_CLIENT_CA_FILE}
);

=head2 key_file

Path to key_file, defaults to $ENV{ETCD_CLIENT_KEY_FILE}

=cut

has key_file => (
    is      => 'ro',
    default => $ENV{ETCD_CLIENT_KEY_FILE}
);

=head2 cert_file

Path to cert_file, defaults to $ENV{ETCD_CLIENT_CERT_FILE}

=cut

has cert_file => (
    is      => 'ro',
    default => $ENV{ETCD_CLIENT_CERT_FILE}
);

=head2 cacert

Path to cacert, defaults to $ENV{ETCD_CLIENT_CACERT_FILE}.

=cut

has cacert => (
    is      => 'ro',
    default => $ENV{ETCD_CLIENT_CACERT_FILE}
);

=head2 ssl

To enable set to 1

=cut

has ssl => (
    is  => 'ro',
    isa => Bool,
);

=head2 api_version

defaults to /v3beta

=cut

has api_version => (
    is      => 'ro',
    isa     => Str,
    default => '/v3beta'
);

=head2 api_path

The full api path. Defaults to http://127.0.0.1:2379/v3alpha

=cut

has api_path => ( is => 'lazy' );

sub _build_api_path {
    my ($self) = @_;
    return ( $self->{ssl}
          || $self->{ca_file}
          || $self->{key_file}
          || $self->{cafile}
          || $self->{cert_file} ? 'https' : 'http' )
      . '://'
      . $self->host . ':'
      . $self->port
      . $self->api_version;
}

=head2 auth_token

The token that is passed during authentication.  This is generated during the
authentication process and stored until no longer valid or username is changed.

=cut

has auth_token => (
    is      => 'rwp',
    clearer => 1,
);

=head1 PUBLIC METHODS

=head2 version

Returns the etcd member version

    $etcd->version()

=cut

sub version {
    my ( $self, $options ) = @_;
    my $cb = pop if ref $_[-1] eq 'CODE';
    my $member = Net::Etcd::Maintenance->new(
        etcd => $self,
        cb   => $cb,
        ( $options ? %$options : () ),
    );
    return $member->version;
}

=head2 watch

See L<Net::Etcd::Watch>

    $etcd->watch({ key =>'foo', range_end => 'fop' })

=cut

sub watch {
    my ( $self, $options ) = @_;
    my $cb = pop if ref $_[-1] eq 'CODE';
    return Net::Etcd::Watch->new(
        etcd => $self,
        cb   => $cb,
        ( $options ? %$options : () ),
    );
}

=head2 role

See L<Net::Etcd::Auth::Role>

    $etcd->role({ role => 'foo' });

=cut

sub role {
    my ( $self, $options ) = @_;
    my $cb = pop if ref $_[-1] eq 'CODE';
    return Net::Etcd::Auth::Role->new(
        etcd => $self,
        cb   => $cb,
        ( $options ? %$options : () ),
    );
}

=head2 role_perm

See L<Net::Etcd::Auth::RolePermission>

Grants or revoke permission of a specified key or range to a specified role.

    $etcd->role_perm(
        { name => 'myrole', key => 'bar', permType => 'READ', prefix => 1 } )->grant;

=cut

sub role_perm {
    my ( $self, $options ) = @_;
    my $cb = pop if ref $_[-1] eq 'CODE';
    my $perm = Net::Etcd::Auth::RolePermission->new(
        etcd => $self,
        cb   => $cb,
        ( $options ? %$options : () ),
    );
}

=head2 user_role

See L<Net::Etcd::User::Role>

    $etcd->user_role({ name => 'samba', role => 'foo' });

=cut

sub user_role {
    my ( $self, $options ) = @_;
    my $cb = pop if ref $_[-1] eq 'CODE';
    return Net::Etcd::User::Role->new(
        etcd => $self,
        cb   => $cb,
        ( $options ? %$options : () ),
    );
}

=head2 auth

See L<Net::Etcd::Auth>

    $etcd->auth({ name => 'samba', password => 'foo' })->authenticate;
    $etcd->auth()->enable;
    $etcd->auth()->disable

=cut

sub auth {
    my ( $self, $options ) = @_;
    my $cb = pop if ref $_[-1] eq 'CODE';
    return Net::Etcd::Auth->new(
        etcd => $self,
        cb   => $cb,
        ( $options ? %$options : () ),
    );
}

=head2 lease

See L<Net::Etcd::Lease>

    $etcd->lease( { ID => 7587821338341002662, TTL => 20 } )->grant;

=cut

sub lease {
    my ( $self, $options ) = @_;
    my $cb = pop if ref $_[-1] eq 'CODE';
    return Net::Etcd::Lease->new(
        etcd => $self,
        cb   => $cb,
        ( $options ? %$options : () ),
    );
}

=head2 maintenance

See L<Net::Etcd::Maintenance>

    $etcd->maintenance()->snapshot

=cut

sub maintenance {
    my ( $self, $options ) = @_;
    my $cb = pop if ref $_[-1] eq 'CODE';
    return Net::Etcd::Maintenance->new(
        etcd => $self,
        cb   => $cb,
        ( $options ? %$options : () ),
    );
}

=head2 member

See L<Net::Etcd::Member>

    $etcd->member()->list

=cut

sub member {
    my ( $self, $options ) = @_;
    my $cb = pop if ref $_[-1] eq 'CODE';
    return Net::Etcd::Member->new(
        etcd => $self,
        cb   => $cb,
        ( $options ? %$options : () ),
    );
}

=head2 user

See L<Net::Etcd::User>

    $etcd->user( { name => 'samba', password => 'foo' } )->add;

=cut

sub user {
    my ( $self, $options ) = @_;
    my $cb = pop if ref $_[-1] eq 'CODE';
    return Net::Etcd::User->new(
        etcd => $self,
        cb   => $cb,
        ( $options ? %$options : () ),
    );
}

=head2 put

See L<Net::Etcd::KV::Put>

    $etcd->put({ key =>'foo1', value => 'bar' });

=cut

=head2 deleterange

See L<Net::Etcd::KV::DeleteRange>

    $etcd->deleterange({ key=>'test0' });

=cut

=head2 range

See L<Net::Etcd::KV::Range>

    $etcd->range({ key =>'test0', range_end => 'test100' });

=cut

=head2 txn

See L<Net::Etcd::KV::Txn>

    $etcd->txn({ compare => \@compare, success => \@op });

=cut

=head2 op

See L<Net::Etcd::KV::Op>

    $etcd->op({ request_put => $put });
    $etcd->op({ request_delete_range => $range });

=cut

=head2 compare

See L<Net::Etcd::KV::Compare>

    $etcd->compare( { key => 'foo', result => 'EQUAL', target => 'VALUE', value => 'baz' });
    $etcd->compare( { key => 'foo', target => 'CREATE', result => 'NOT_EQUAL', create_revision => '2' });

=cut

=head2 configuration

Initialize configuration checks to see if etcd is installed locally.

=cut

sub configuration {
    Net::Etcd::Config->configuration;
}

=head1 AUTHOR

Sam Batschelet (hexfusion)

=head1 CONTRIBUTORS

Ananth Kavuri
Michael Fung

=head1 ACKNOWLEDGEMENTS

The L<etcd|https://github.com/etcd-io/etcd> developers and community.

=head1 CAVEATS

The L<etcd|https://github.com/etcd-io/etcd> v3 API is in heavy development and can change at anytime please see
 L<api_reference_v3|https://github.com/etcd-io/etcd/blob/master/Documentation/dev-guide/api_reference_v3.md>
for latest details.

Authentication provided by this module will only work with etcd v3.3.0+

=head1 LICENSE AND COPYRIGHT

Copyright 2018 Sam Batschelet (hexfusion).

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

1;
